### ISSUE TITLE
Reproducing the pycocoeval metrics (BLEU-4, SPICE, etc.)

### ISSUE THREAD

**Issue by jmhessel** [ck0123/improved-bertscore-for-image-captioning-evaluation#3]
Hi There!

Thanks for the awesome paper and code release! I was wondering if you could possibly make available the code you used to reproduce the correlations in Table 3. For the flickr8k json you provided here containing the expert annotations, I tried computing them myself. [I used the script in this gist](https://gist.github.com/jmhessel/7939a6fe28120b9ed5275260a0f31d30). I've double checked that the BLEU scoring matches the outputs of both NLTK and sacrebleu. It is computed using the [python3 fork of pycocoevalcap](https://github.com/salaniz/pycocoevalcap)

If you'd like to run it yourself, you can use the flickr8k json in this repo as:
```
python reproduce.py improved-bertscore-for-image-captioning-evaluation/example/flickr.json
```
The result is significantly different than table 3 (particularly for spice!), which makes me think that I have something setup wrong.
```
Metric spice, kendall=0.562 (vs. .458 in the paper)
Metric bleu, kendall=0.321 (vs .140 in the paper)
Metric meteor, kendall=0.450 (vs. .436 in the paper)
Metric rouge, kendall=0.336 (vs. .323 in the paper)
Metric cider, kendall=0.468 (vs. .447 in the paper)
```

What do you think? And thanks again!!

Jack

**Comment by ck0123** [comment]
Hi Jack,
Sorry for late reply.
Since I has been already graduated and my old Macbook has been "retired", it is hard for me to check what's the difference in detail between your gist and my code. 
Something I can remember is that: in my experiment, firstly I can reproduce SPICE paper Flickr 8K experiments(only that I confirmed my code was correct) . I have checked that SPICE is never update, pcocoeval maybe not update too. A possible target may Standford NLP tools, since at that time I used 3.6.0 as  recommended by SPICE but current version is 4.2.0. Maybe the update of Standford NLP tools helps SPICE a lot? But still, I can't explain why BLEU-4 is 0.32(it is too high indeed).  Another paper called TIGEr also report BLEU-4 in Flickr 0.1x.

BEST,
Yanzhi

**Comment by jmhessel** [comment]
Hi Yanzhi!

Thanks for the quick reply, and for your thoughts regarding my setup! And, thanks for the reference re: TIGEr. Even the original spice paper reports a much lower BLEU correlation, so it's very possible I'm missing something.

For BLEU, I did check to see if the pycocoeval cap BLEU matches nltk and sacrebleu:
```
Sacrebleu: 4.147909, pycocoevalcap: 4.147909, nltk: 4.14666
```
so that looks okay...

regarding spice and CoreNLP updates, I suspected that too... I did check to see if the pycocoeval cap outputs matched the outputs from the original spice repo, with an example json they provided.

https://github.com/peteanderson80/SPICE (which downloads CoreNLP 3.6.0)

```
> java -Xmx8G -jar spice-*.jar SPICE/example_input_larger.json
********  SPICE Evaluation  ********

All tuples
  f-score:	0.133 (SPICE metric)
  precision:	0.330
  recall:	0.086
  true pos:	1935
  false pos:	4114
  false neg:	21281
  num images:	1000
```

And pycocoevalcap gives:
```
spice 0.133
```
Which checks out... Evaluation standardization is so tricky! I'll do some more thinking re: spice evaluation :-)

Thanks again!

Jack
